src/MediaWiki.ts

```diff
@@ -28,35 +30,38 @@ class MediaWiki {
   private readonly password: string
   private readonly apiPath: string
   private readonly domain: string
-  private readonly articleApiUrlBase: string
+  private apiUrlDirector: ApiURLDirector
 
   constructor(config: MWConfig) {
     this.domain = config.domain || ''
     this.username = config.username
     this.password = config.password
     this.getCategories = config.getCategories
 
-    this.baseUrl = new URL(ensureTrailingChar(config.base, '/'))
+    this.baseUrl = basicURLDirector.buildMediawikiBaseURL(config.base)
 
     this.apiPath = config.apiPath ?? 'w/api.php'
     this.wikiPath = config.wikiPath ?? DEFAULT_WIKI_PATH
 
-    this.webUrl = new URL(this.wikiPath, this.baseUrl)
-    this.apiUrl = new URL(`${this.apiPath}?`, this.baseUrl)
+    const baseUrlDirector = new BaseURLDirector(this.baseUrl.href)
 
-    this.veApiUrl = new URL(`${this.apiUrl.href}action=visualeditor&mobileformat=html&format=json&paction=parse&page=`)
+    this.webUrl = baseUrlDirector.buildURL(this.wikiPath)
+    this.apiUrl = baseUrlDirector.buildURL(this.apiPath)
 
-    this.restApiUrl = new URL(ensureTrailingChar(new URL(config.restApiPath ?? 'api/rest_v1', this.baseUrl.href).toString(), '/'))
-    this.mobileRestApiUrl = new URL(ensureTrailingChar(new URL(config.restApiPath ?? 'api/rest_v1/page/mobile-sections', this.baseUrl.href).toString(), '/'))
-    this.desktopRestApiUrl = new URL(ensureTrailingChar(new URL(config.restApiPath ?? 'api/rest_v1/page/html', this.baseUrl.href).toString(), '/'))
+    this.apiUrlDirector = new ApiURLDirector(this.apiUrl.href)
 
-    this.modulePath = `${urlParser.resolve(this.baseUrl.href, config.modulePath ?? 'w/load.php')}?`
-    this.articleApiUrlBase = `${this.apiUrl.href}action=parse&format=json&prop=${encodeURI('modules|jsconfigvars|headhtml')}&page=`
+    this.veApiUrl = this.apiUrlDirector.buildVisualEditorURL()
+
+    this.restApiUrl = baseUrlDirector.buildRestApiURL(config.restApiPath)
+    this.mobileRestApiUrl = baseUrlDirector.buildMobileRestApiURL(config.restApiPath)
+    this.desktopRestApiUrl = baseUrlDirector.buildDesktopRestApiURL(config.restApiPath)
```

Comment:
This is the URL which should be used for the MainPage download (special case)... that you have to redefine it somewhere elase sounds weird to me (see comment above). You really need to be clear where all this URL handling should take place IMO, in Mediawiki or in Downloader (this is why scoping in a banner comment for both class might help to better segregate each duty).

Reply:
> This is the URL which should be used for the MainPage download (special case)
It will be used